libvast/src/format/csv.cpp

```diff
@@ -129,4 +139,265 @@ const char* writer::name() const {
+                              defaults::kvp_separator);
+}
+
+void reader::reset(std::unique_ptr<std::istream> in) {
+  VAST_ASSERT(in != nullptr);
+  input_ = std::move(in);
+  lines_ = std::make_unique<detail::line_range>(*input_);
+}
+
+caf::error reader::schema(vast::schema s) {
+  schema_ = std::move(s);
+  for (auto& t : s) {
+    if (auto r = caf::get_if<record_type>(&t))
+      schema_.add(flatten(*r));
+  }
+  return caf::none;
+}
+
+vast::schema reader::schema() const {
+  return schema_;
+}
+
+const char* reader::name() const {
+  return "csv-reader";
+}
+
+caf::optional<record_type>
+reader::make_layout(const std::vector<std::string>& names) {
+  for (auto& t : schema_) {
+    if (auto r = caf::get_if<record_type>(&t)) {
+      auto select_fields = [&]() -> caf::optional<record_type> {
+        std::vector<record_field> result_raw;
+        for (auto& name : names) {
+          if (auto field = r->at(name))
+            result_raw.emplace_back(name, *field);
+          else
+            return caf::none;
+        }
+        return record_type{std::move(result_raw)};
+      };
+      if (auto result = select_fields())
+        return result->name(r->name());
+    } else if (names.size() == 1 && names[0] == t.name()) {
+      // Hoist naked type into record.
+      // TODO: Maybe this is actually not a good idea?
+      return record_type{{t.name(), t}}.name(t.name());
+    } // else skip
+  }
+  return caf::none;
+}
+
+namespace {
+
+template <class Iterator, class Attribute>
+struct container_parser_builder {
+  // In case we run into performance issue with the parsers
+  // generated here, `rule` could be replaced by `type_erased_parser`
+  // to eliminated one indirection. This requires attribute support
+  // in `type_erased_parser`.
+  using result_type = rule<Iterator, Attribute>;
+
+  explicit container_parser_builder(options opt) : opt_{std::move(opt)} {
+    // nop
+  }
+
+  // TODO: support enumeration_type inside of containers too.
+  template <class T>
+  result_type operator()(const T& t) const {
+    if constexpr (std::is_same_v<T, string_type>) {
+      // clang-format off
+      return +(parsers::any - opt_.set_separator - opt_.kvp_separator) ->* [](std::string x) {
+        return data{std::move(x)};
+      };
+    } else if constexpr (std::is_same_v<T, pattern_type>) {
+      return +(parsers::any - opt_.set_separator - opt_.kvp_separator) ->* [](std::string x) {
+        return data{pattern{std::move(x)}};
+      };
+      // clang-format on
+    } else if constexpr (std::is_same_v<T, enumeration_type>) {
+      auto to_enumeration = [t](std::string s) -> caf::optional<Attribute> {
+        auto i = std::find(t.fields.begin(), t.fields.end(), s);
+        if (i == t.fields.end()) {
+          VAST_WARNING_ANON("csv reader failed to parse unexpected enum value",
+                            s);
+          return caf::none;
+        }
+        return enumeration(std::distance(t.fields.begin(), i));
```

Comment:
As far as I understand, the enums are currently not handled at the query side. The binding can only be done "late" anyway, because the layout is needed to map it to the internal representation.

Extensibility seems to be orthogonal to me. One way that I can come up with now is to maintain an internal enum defnition that sorts the possible values by time of declaration. So if an enum is read in a layout, instead of just putting it into the record verbatim, we look it up by its fqn in an enums table and update the entry there by appending any new values from the parsed definition.

Reply:
> [..] because the layout is needed to map it to the internal representation.

Okay, the layout is usually around when we're dealing with enum values, since they are inside a table slice. That means for a given enum type, we can create a specific value index. But since the numeric value is an implementation detail, and users would actually query by enum string value, it won't be a problem.